### PR TITLE
Kconfig: add CONFIG_SND_SOC_SOF_HDA_AUDIO_CODEC dependent

### PR DESCRIPTION
**Pull request by libinyang**
If the driver uses common hdmi, SND_SOC_SOF_HDA_AUDIO_CODEC is required  to be enabled. Otherwise, the legacy hda codec driver will not be used. This patchset add the SND_SOC_SOF_HDA_AUDIO_CODEC dependent and remove the ifdef in the source code

Fixes https://github.com/thesofproject/linux/issues/1968

### CODE REVIEW

File: sound/soc/intel/boards/Kconfig [hunk 1]
```diff
@@ -304,7 +304,7 @@ config SND_SOC_INTEL_BXT_DA7219_MAX98357A_MACH
 	tristate "Broxton with DA7219 and MAX98357A in I2S Mode"
 	depends on I2C && ACPI
 	depends on MFD_INTEL_LPSS || COMPILE_TEST
-	depends on SND_HDA_CODEC_HDMI
+	depends on SND_HDA_CODEC_HDMI && SND_SOC_SOF_HDA_AUDIO_CODEC
```
Comment (kv2019i) [inline sound/soc/intel/boards/Kconfig]: @libinyang Actually this is a mistake in my earlier patch. This mach driver is not used with SOF, so it shouldn't have SND_HDA_CODEC_HDMI.

File: sound/soc/intel/boards/Kconfig [hunk 2]
```diff
@@ -316,6 +316,7 @@ config SND_SOC_INTEL_BXT_RT298_MACH
 	tristate "Broxton with RT298 I2S mode"
 	depends on I2C && ACPI
 	depends on MFD_INTEL_LPSS || COMPILE_TEST
+	depends on SND_HDA_CODEC_HDMI && SND_SOC_SOF_HDA_AUDIO_CODEC
```
Comment (kv2019i) [inline sound/soc/intel/boards/Kconfig]: same here

File: sound/soc/intel/boards/Kconfig [hunk 3]
```diff
@@ -419,7 +420,7 @@ config SND_SOC_INTEL_GLK_DA7219_MAX98357A_MACH
 	tristate "GLK with DA7219 and MAX98357A in I2S Mode"
 	depends on I2C && ACPI
 	depends on MFD_INTEL_LPSS || COMPILE_TEST
-	depends on SND_HDA_CODEC_HDMI
+	depends on SND_HDA_CODEC_HDMI && SND_SOC_SOF_HDA_AUDIO_CODEC
```
Comment (kv2019i) [inline sound/soc/intel/boards/Kconfig:422]: Here I'd just change the surrounding ifdef from SND_SOC_SOF_HDA_LINK to SND_SOC_SOF_HDA_AUDIO_CODEC. If all the GLk machine drivers need SOF and HDA codec drivers, then why not just have it in the ifdef.

File: sound/soc/intel/boards/Kconfig [hunk 4]
```diff
@@ -431,7 +432,7 @@ config SND_SOC_INTEL_GLK_RT5682_MAX98357A_MACH
 	tristate "GLK with RT5682 and MAX98357A in I2S Mode"
 	depends on I2C && ACPI
 	depends on MFD_INTEL_LPSS || COMPILE_TEST
-	depends on SND_HDA_CODEC_HDMI
+	depends on SND_HDA_CODEC_HDMI && SND_SOC_SOF_HDA_AUDIO_CODEC
```
Comment (kv2019i) [inline sound/soc/intel/boards/Kconfig:434]: same as above

File: sound/soc/intel/boards/Kconfig [hunk 5]
```diff
@@ -448,7 +449,7 @@ if SND_SOC_INTEL_SKYLAKE_HDAUDIO_CODEC || SND_SOC_SOF_HDA_AUDIO_CODEC
 
 config SND_SOC_INTEL_SKL_HDA_DSP_GENERIC_MACH
 	tristate "SKL/KBL/BXT/APL with HDA Codecs"
-	depends on SND_HDA_CODEC_HDMI
+	depends on SND_HDA_CODEC_HDMI && SND_SOC_SOF_HDA_AUDIO_CODEC
```
Comment (kv2019i) [inline sound/soc/intel/boards/Kconfig]: This is not needed as SND_SOC_SOF_HDA_AUDIO_CODEC is already in the "if" statement above.

File: sound/soc/intel/boards/Kconfig [hunk 6]
```diff
@@ -466,7 +467,7 @@ config SND_SOC_INTEL_SOF_RT5682_MACH
 	depends on I2C && ACPI
 	depends on (SND_SOC_SOF_HDA_LINK && (MFD_INTEL_LPSS || COMPILE_TEST)) ||\
 		   (SND_SOC_SOF_BAYTRAIL && (X86_INTEL_LPSS || COMPILE_TEST))
-	depends on SND_HDA_CODEC_HDMI
+	depends on SND_HDA_CODEC_HDMI && SND_SOC_SOF_HDA_AUDIO_CODEC
```
Comment (kv2019i) [inline sound/soc/intel/boards/Kconfig]: @libinyang Here I'd merge this with SOF_HDA_LINK on L468. If this mach driver needs the audio codecs, then no need to test with HDA_LINK here.

File: sound/soc/intel/boards/Kconfig [hunk 7]
```diff
@@ -510,7 +511,7 @@ config SND_SOC_INTEL_SOF_CML_RT1011_RT5682_MACH
 	tristate "CML with RT1011 and RT5682 in I2S Mode"
 	depends on I2C && ACPI
 	depends on MFD_INTEL_LPSS || COMPILE_TEST
-	depends on SND_HDA_CODEC_HDMI
+	depends on SND_HDA_CODEC_HDMI && SND_SOC_SOF_HDA_AUDIO_CODEC
```
Comment (kv2019i) [inline sound/soc/intel/boards/Kconfig:514]: same as above

File: sound/soc/intel/boards/Kconfig [hunk 8]
```diff
@@ -529,7 +530,7 @@ config SND_SOC_INTEL_SOF_DA7219_MAX98373_MACH
 	tristate "SOF with DA7219 and MAX98373/MAX98360A in I2S Mode"
 	depends on I2C && ACPI
 	depends on MFD_INTEL_LPSS || COMPILE_TEST
-	depends on SND_HDA_CODEC_HDMI
+	depends on SND_HDA_CODEC_HDMI && SND_SOC_SOF_HDA_AUDIO_CODEC
```
Comment (kv2019i) [inline sound/soc/intel/boards/Kconfig:533]: This is ok here, but to align with with other platforms, could be put as additional condition for "if" at L527

File: sound/soc/intel/boards/Kconfig [hunk 9]
```diff
@@ -547,7 +548,7 @@ config SND_SOC_INTEL_EHL_RT5660_MACH
 	tristate "EHL with RT5660 in I2S mode"
 	depends on I2C && ACPI
 	depends on MFD_INTEL_LPSS || COMPILE_TEST
-	depends on SND_HDA_CODEC_HDMI
+	depends on SND_HDA_CODEC_HDMI && SND_SOC_SOF_HDA_AUDIO_CODEC
```
Comment (kv2019i) [inline sound/soc/intel/boards/Kconfig:551]: Same comment as for JSL change.
This not in Mark's tree either, so can you put these in a separate patch?

File: sound/soc/intel/boards/Kconfig [hunk 10]
```diff
@@ -564,7 +565,7 @@ config SND_SOC_INTEL_SOUNDWIRE_SOF_MACH
 	depends on MFD_INTEL_LPSS || COMPILE_TEST
 	depends on SND_SOC_INTEL_USER_FRIENDLY_LONG_NAMES || COMPILE_TEST
 	depends on SOUNDWIRE
-	depends on SND_HDA_CODEC_HDMI
+	depends on SND_HDA_CODEC_HDMI && SND_SOC_SOF_HDA_AUDIO_CODEC
```
Comment (kv2019i) [inline sound/soc/intel/boards/Kconfig:568]: This is ok.

File: sound/soc/intel/boards/Kconfig (hunk 1 above)
Comment (libinyang) [reply sound/soc/intel/boards/Kconfig]: @kv2019i Do you mean we should remove `SND_HDA_CODEC_HDMI` in the Kconfig? I checked and found soc-acpi-intel-bxt-match.c has already added `.sof_fw_filename = "sof-apl.ri"` and `.sof_tplg_filename = "sof-apl-da7219.tplg"` in `snd_soc_acpi_intel_bxt_machines`. Does this mean our SOF has already supported it? Also I didn't find the driver will set it to sst driver in "config_table"
If we remove `SND_HDA_CODEC_HDMI`, I think we should also remove some code related to the common hdmi.

File: sound/soc/intel/boards/Kconfig (hunk 3 above)
Comment (libinyang) [reply sound/soc/intel/boards/Kconfig:422]: OK

File: sound/soc/intel/boards/Kconfig (hunk 6 above)
Comment (libinyang) [reply sound/soc/intel/boards/Kconfig]: @kv2019i OK, I will move SND_HDA_CODEC_HDMI && SND_SOC_SOF_HDA_AUDIO_CODEC to L468

File: sound/soc/intel/boards/Kconfig (hunk 5 above)
Comment (libinyang) [reply sound/soc/intel/boards/Kconfig]: @kv2019i In the "if" statement, it requires "if SND_SOC_INTEL_SKYLAKE_HDAUDIO_CODEC || SND_SOC_SOF_HDA_AUDIO_CODEC"
This means SND_SOC_SOF_HDA_AUDIO_CODEC is not a "must". Let's look into the first condition "SND_SOC_INTEL_SKYLAKE_HDAUDIO_CODEC". In its description, "This option is DEPRECATED". And I checked the src code, This config is only used in SST driver now. So I believe the "if" statement is expired. We should just use "if SND_SOC_SOF_HDA_AUDIO_CODEC".

File: sound/soc/intel/boards/Kconfig (hunk 7 above)
Comment (libinyang) [reply sound/soc/intel/boards/Kconfig:514]: @kv2019i Do you mean the same as "https://github.com/thesofproject/linux/pull/1989#discussion_r404732783". In this config, I would prefer "depends on SND_HDA_CODEC_HDMI && SND_SOC_SOF_HDA_AUDIO_CODEC" to “depends on SND_SOC_SOF_HDA_AUDIO_CODEC && (MFD_INTEL_LPSS || COMPILE_TEST)”. This is because it needs "SND_SOC_SOF_HDA_AUDIO_CODEC" is enabled only because we are using SND_HDA_CODEC_HDMI. So it's better we put them together

File: sound/soc/intel/boards/Kconfig (hunk 8 above)
Comment (libinyang) [reply sound/soc/intel/boards/Kconfig:533]: @kv2019i As above, I would like to use "depends on SND_HDA_CODEC_HDMI && SND_SOC_SOF_HDA_AUDIO_CODEC".  SND_HDA_CODEC_HDMI and SND_SOC_SOF_HDA_AUDIO_CODEC is the combined condition to enable common hdmi. If we put it in the "if" statement, the relationship between SND_HDA_CODEC_HDMI and SND_SOC_SOF_HDA_AUDIO_CODEC is not so clear to user.

File: sound/soc/intel/boards/Kconfig (hunk 9 above)
Comment (libinyang) [reply sound/soc/intel/boards/Kconfig:551]: OK. I will make a separate patch for it. And for the "if" statement, please see my previous comments.

File: sound/soc/intel/boards/Kconfig (hunk 1 above)
Comment (kv2019i) [reply sound/soc/intel/boards/Kconfig]: @libinyang Good question. I saw Pierre actually removed SOF support for this BXT driver with commit "ASoC: Intel: boards: remove support for SOF for APL devices" (this is after my common-hdmi series). So I would not touch this machine driver in this patch.

@plbossart Can you recall what is the history? We probably should have removed SND_HDA_CODEC_HDMI depends when we removed SOF support.

File: sound/soc/intel/boards/Kconfig (hunk 5 above)
Comment (kv2019i) [reply sound/soc/intel/boards/Kconfig]: Discussion ongoing on that deprecation, we can't remove it now. I think the if check is still right. If SST is used (SKYLAKE_HD_AUDIO_CODEC part of the if check), the machine driver will not use common HDMI driver but will use hdac-hdmi. If SOF is used (the other part of the if), the machine driver will always set up common codec and SND_HDA_CODEC_HDMI is needed.

For simplicity, SND_HDA_CODEC_HDMI is selected always, but to be really accurate, it should only be a dependency, if SOF is enabled in the build.

I don't think you need to modify this.

File: sound/soc/intel/boards/Kconfig (hunk 7 above)
Comment (kv2019i) [reply sound/soc/intel/boards/Kconfig:514]: @libinyang Fair enough, for I2S machine drivers this makes sense indeed.

File: sound/soc/intel/boards/Kconfig (hunk 8 above)
Comment (kv2019i) [reply sound/soc/intel/boards/Kconfig:533]: Ack.

File: sound/soc/intel/boards/Kconfig (hunk 9 above)
Comment (kv2019i) [reply sound/soc/intel/boards/Kconfig:551]: Wait a sec, I just reviewed #1996 from Pierre and Pierre has put his changes in one patch, so maybe this is needed. Then this just means we need to send the machine driver first, and only then can send this fix, but should be ok.

File: sound/soc/intel/boards/Kconfig (hunk 1 above)
Comment (libinyang) [reply sound/soc/intel/boards/Kconfig]: @kv2019i Got it. I will remove my fix for this machine driver. Thanks.

File: sound/soc/intel/boards/Kconfig (hunk 9 above)
Comment (libinyang) [reply sound/soc/intel/boards/Kconfig:551]: @kv2019i Both are OK to me : ) I will make it in the same patch.

File: sound/soc/intel/boards/Kconfig (hunk 5 above)
Comment (libinyang) [reply sound/soc/intel/boards/Kconfig]: @kv2019i If so, I think we have 2 choice:
1. use "if SND_SOC_INTEL_SKYLAKE_HDAUDIO_CODEC || (SND_SOC_SOF_HDA_AUDIO_CODEC && SND_HDA_CODEC_HDMI)" in the if statement, or
2. we don't change the if statement and use 'depends on SND_HDA_CODEC_HDMI && SND_SOC_SOF_HDA_AUDIO_CODEC' which is used in my original patch.
Or do you have other suggestions?

For the first choice, I still have a tiny concern. It is possible that SND_SOC_INTEL_SKYLAKE_HDAUDIO_CODEC is enabled and (SND_SOC_SOF_HDA_AUDIO_CODEC && SND_HDA_CODEC_HDMI) are not both enabled. 
So in this case, it can be: SND_SOC_SOF_HDA_AUDIO_CODEC is disabled and SND_HDA_CODEC_HDMI is enabled, and SND_SOC_INTEL_SKYLAKE_HDAUDIO_CODEC is enabled while we are still using SOF driver. This will be wrong.

I'm not sure if this option is acceptable: let's remove the 'if' statement and use:
depends on SND_SOC_SOF_HDA_AUDIO_CODEC
depends on SND_HDA_CODEC_HDMI

File: sound/soc/intel/boards/Kconfig (hunk 1 above)
Comment (plbossart) [reply sound/soc/intel/boards/Kconfig]: This machine driver is used for APL Chromebooks. We cannot change the firmware key to use the SOF public one except on preproduction hardware. Seppo and I are the only ones to have such devices. 

Now that the firmware is signed with the production key, we could add SOF support here. I am not sure if we want to, or this would be something to discuss separately. Let's leave it as is for now.

**Comment by libinyang** [comment]
patches are updated based on @kv2019i review. 
Only "SND_SOC_INTEL_SKL_HDA_DSP_GENERIC_MACH" section is still open. And I will update the patches based on the final solution.

File: sound/soc/intel/boards/Kconfig (hunk 1 above)
Comment (libinyang) [reply sound/soc/intel/boards/Kconfig]: @plbossart Thanks for the information. I will not touch SND_SOC_INTEL_BXT_DA7219_MAX98357A_MACH and SND_SOC_INTEL_BXT_RT298_MACH in this patch.

File: sound/soc/intel/boards/Kconfig (hunk 5 above)
Comment (kv2019i) [reply sound/soc/intel/boards/Kconfig]: @libinyang I'd vote for option

3. don't change the definitions at all, I think this is ok for SST and SOF

I think only problem is that SND_HDA_CODEC_HDMI needs to be selected to compile for SST, but all distributions have SND_HDA_CODEC_HDMI enabled (due to legacy), so this seems a pretty minor problem. Code will still work, worst that can happen is that you can't select SST.

For the other options, I think the drawbacks are bigger:
1. I think you are right that iIf SOC_INTEL_SKYLAKE_HDAUDIO_CODEC and SOF_HDA_AUDIO_CODEC are defined but not HDA_CODEC_HDMI, and you use this on a system with dmics, you will hit an error at runtime as HDMI probe will fail. This seems much worse than the drawback of option 3.

2. This means that to build SST SKL driver, you'd need to also enable SOF in kernel config. This seems a bigger problem than (3) as SST and SOF are not typically enabled in distro configs by default and it is very likely if you are enabling SST driver, you will miss some SOF options.

I still think (3) is least bad.

**Comment by libinyang** [comment]
@kv2019i Fair enough. I won't touch SKL hda generic driver config this time. It seems there isn't a perfect solution because of the limitation of the kconfig.

File: sound/soc/intel/boards/Kconfig [hunk 11]
```diff
@@ -413,7 +413,7 @@ config SND_SOC_INTEL_KBL_RT5660_MACH
 
 endif ## SND_SOC_INTEL_KBL
 
-if SND_SOC_SOF_GEMINILAKE  && SND_SOC_SOF_HDA_LINK
+if SND_SOC_SOF_GEMINILAKE  && SND_SOC_SOF_HDA_AUDIO_CODEC
```
Comment (plbossart) [inline sound/soc/intel/boards/Kconfig]: probably best if you move the dependency below, so that all users of HDMI rely on the same pattern of 
	depends on SND_HDA_CODEC_HDMI && SND_SOC_SOF_HDA_AUDIO_CODEC

Comment (libinyang) [reply sound/soc/intel/boards/Kconfig]: @plbossart OK. I will remove "SND_SOC_SOF_HDA_AUDIO_CODEC" in the if statement and use "depends on SND_HDA_CODEC_HDMI && SND_SOC_SOF_HDA_AUDIO_CODEC"

File: sound/soc/intel/boards/sof_sdw.c [hunk 12]
```diff
@@ -663,9 +663,7 @@ static int sof_card_dai_links_create(struct device *dev,
 				     struct snd_soc_card *card)
 {
 	int ssp_num, sdw_be_num = 0, hdmi_num = 0, dmic_num;
-#if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA_AUDIO_CODEC)
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_sdw.c:666]: minor nit on the commit message: when you wrote 'judgement' you meant 'test'.

Comment (libinyang) [reply sound/soc/intel/boards/sof_sdw.c:666]: Thanks for pointing it out. I will refine the patch. Sorry for the delay. I missed your comments.

**Comment by plbossart** [comment]
@xiulipan @fredoh9 @wenqingfu CI is stuck on this one too

**Comment by kv2019i** [comment]
Thanks @libinyang , CI now good, merging!
